### PR TITLE
Object-oriented API

### PR DESCRIPTION
**Pull request by TChatzigiannakis**
A partial reconstruction of the C++ API in a manner that feels like native C#, plus tests.

### CODE REVIEW

**Comment by mjsabby** [comment]
@TChatzigiannakis I've started looking at this now.

File: LLVMSharp.sln [hunk 1]
```diff
@@ -1,20 +1,46 @@
 ﻿
 Microsoft Visual Studio Solution File, Format Version 12.00
 # Visual Studio 15
-VisualStudioVersion = 15.0.26730.15
+VisualStudioVersion = 15.0.26730.16
 MinimumVisualStudioVersion = 10.0.40219.1
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "LLVMSharp", "src\LLVMSharp.csproj", "{3EA66A6D-9DC3-46FF-AA7F-D0BB5BF0BA73}"
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "LLVMSharp", "src\LLVMSharp.csproj", "{3EA66A6D-9DC3-46FF-AA7F-D0BB5BF0BA73}"
+EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Tests", "Tests\Tests.csproj", "{48FED920-3F0B-4BCB-A3AA-C9F31DAA385B}"
```
Comment (am11) [inline LLVMSharp.sln]: `Tests\Tests.csproj` -> `tests\Tests.csproj` to satisfy the case-sensitive OS: https://travis-ci.org/Microsoft/LLVMSharp/builds/419398392#L565

File: LLVMSharp.sln [hunk 2]
```diff
@@ -1,20 +1,46 @@
...
 EndProject
 Global
 	GlobalSection(SolutionConfigurationPlatforms) = preSolution
 		Debug|Any CPU = Debug|Any CPU
+		Debug|x64 = Debug|x64
+		Debug|x86 = Debug|x86
```
Comment (am11) [inline LLVMSharp.sln]: Additional platforms get added to SLN file when we add a new project. However, LLVMSharp is AnyCPU by design. To remove these from VS:

image

then click save-all (ctrl+shift+s).

**Comment by TChatzigiannakis** [comment]
Travis CI build fails because a .NET Framework 4.6.1 test project has been added.

If someone wants the test project to be .NET Core compatible, you are welcome to do so. I'm still very unfamiliar with .NET Core and I don't know how to make `dotnet test` properly locate the tests.

File: LLVMSharp.sln (hunk 1 above)
Comment (TChatzigiannakis) [reply LLVMSharp.sln]: Thank you for pointing this out. Unfortunately, it's a bit more complicated than that — some files were in `tests` and some were in `Tests`. Windows kept putting all files in one directory whenever I cloned, so I ended up using the GitHub web interface to manually move those files.

File: LLVMSharp.sln (hunk 2 above)
Comment (TChatzigiannakis) [reply LLVMSharp.sln]: Thank you for your input.

Back when I added those targets in the branch, it was the case that tests weren't properly discovered within Visual Studio when the AnyCPU configuration was selected. Clearly that's not the case anymore, so I've removed them.

**Comment by TChatzigiannakis** [comment]
Do we have any news progress on the review of this PR?

**Comment by mjsabby** [comment]
@TChatzigiannakis things mostly look good, but can you please create a single commit here? that is branch of latest master. I will then after checking in your PR bring in LLVM 7 stuff.

**Pull request closed without merging**